### PR TITLE
New Improved ceph admin modules with support of flexible orchestration options

### LINKED ISSUE #359: remove "chmod 777 /etc/ceph" for cephadm
[ceph/ceph_admin/bootstrap_mixin.py](https://github.com/red-hat-storage/cephci/blob/4513932a193de58909851a00cf19fd028e9f7224/ceph/ceph_admin/bootstrap_mixin.py) always runs `chmod 777 /etc/ceph`

Do we need to make that directory world-writable for some reason?

If we could remove this, it would make the test closer to the documented procedures for users, and it would reduce test execution time.

### PR DESCRIPTION
**Pull request by sunilkumarn417**
New design which adds different deployment provision and enhances flexibility.
- added provision for apply label and different placement option.
- supports cephadm orchestration command execution with more CLI arguments.
- updated sanity and deployment test suite as per the new design.
- added remove service provision.
- support to get json/yaml and other supported format CLI response.

This design brings flexibility to QE by supporting staged deployment. Starting with a small number of nodes and then
expanding the cluster. Below illustrates an example of it's strength.

```
tests:
  ...
  - test:
      name: cluster deployment
      desc: Execute the cluster deployment workflow.
      module: test_deployment.py
      polarion-id:
      config:
        steps:
          - config:
              command: bootstrap
              service: cephadm
              args:
                mon-ip: node1
          - config:
              command: add_hosts
              service: host
              args:
                nodes: [node1, node2, node3]
                attach_ip_address: true
                labels: apply-all-labels
          - config:
              command: apply
              service: osd
              args:
                all-available-devices: true
      destroy-cluster: false
      abort-on-fail: true
  - test:
      desc: Perform base line operations
      ...
  ...
  - test:
      name: adding 2 more hosts
      desc: Adding node[4,5]
      module: test_development.py
      config:
        steps:
          - config:
               command: add_hosts
                service: host
                args:
                  nodes: [node3, node4]
                  attach_ip_address: true
                  labels: apply-all-labels
          -  config:
              command: apply
              service: osd
              args:
                 all-available-devices: true
                 unmanaged: true
      destroy-cluster: false
      abort-on-fail: true
  - test
  ...
```

**Test results:**
http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-1613639908753/result.html

Closes #359

### CODE REVIEW

**Comment by rakeshgm** [comment]
phew,  this is a big one :)

**Comment by sunilkumarn417** [comment]
Latest results: http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-1613649237345

- Fixed remote_file issue.

**Comment by psathyan** [comment]
> phew, this is a big one :)

Yes, we couldn't break into small pieces as the required was to stand up a minimal cluster for the FGs.

File: ceph/ceph_admin/apply.py [hunk 1]
```diff
@@ -1,85 +1,122 @@
...
-from ceph.ceph import ResourcesNotFoundError
+from ceph.utils import get_nodes_by_id
 
-logger = logging.getLogger(__name__)
+from .common import config_dict_to_string
+from .typing_ import ServiceProtocol
 
 
-class ServiceRemoveFailure(Exception):
+class ServiceApplyFailure(Exception):
```
Comment (psathyan) [inline ceph/ceph_admin/apply.py:18]: IMO, the name of the exception class could be `OrchApplyFailed`.

File: ceph/ceph_admin/apply.py [hunk 2]
```diff
@@ -1,85 +1,122 @@
 """
+Represent the ceph orch CLI action 'apply'.
+
 Module to deploy ceph role service(s) using orchestration command
 "ceph orch apply <role> [options] --placement '<placements>' "
 
-this module inherited where service deployed using "apply" operation.
-and also validation using orchestration process list response
+This module inherited where service deployed using "apply" operation and also
```
Comment (psathyan) [inline ceph/ceph_admin/apply.py:7]: I believe we could reword here.
```
This is a mixin object and can be applied to the supported service classes.
```

File: ceph/ceph_admin/apply.py [hunk 3]
```diff
@@ -1,85 +1,122 @@
...
+class ServiceApplyFailure(Exception):
     pass
 
 
-class Apply:
-
-    apply_cmd = ["ceph", "orch", "apply"]
-    remove_cmd = ["ceph", "orch", "rm"]
+class ApplyMixin:
+    """Add apply support to the base class."""
```
Comment (psathyan) [inline ceph/ceph_admin/apply.py:23]: I believe it should be `Add apply command support to child object.`

File: ceph/ceph_admin/bootstrap.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        )
+        cmd += custom_image_args.format(
+            user=cdn_cred.get("username"),
+            password=cdn_cred.get("password"),
+        )
+
+        # To be generic, the mon-ip contains the global node name. Here, we replace the
+        # name with the IP address. The replacement allows us to be inline with the
+        # CLI option.
+        mon_node = args.pop("mon-ip")
```
Comment (psathyan) [inline ceph/ceph_admin/bootstrap.py:70]: Should we allow it be optional and take the `installer-node ip address` if the value is not provided?

File: ceph/ceph_admin/host.py [hunk 5]
```diff
@@ -0,0 +1,386 @@
...
+            label_set = set(ceph_node.role.role_list)
+            _labels = list(label_set)
+
+        cmd.extend(["host", "add", ceph_node.shortname])
+
+        if attach_address:
+            cmd.append(ceph_node.ip_address)
+
+        if _labels:
+            # To fill mandate <address> argument, In case attach_address is False
```
Comment (psathyan) [inline ceph/ceph_admin/host.py:82]: Shall we raise a usability issue in the upstream. Having positional arguments as empty does not make sense.

File: ceph/ceph_admin/host.py [hunk 6]
```diff
@@ -0,0 +1,386 @@
...
+            if not attach_address:
+                cmd.append("''")
+
+            cmd += _labels
+
+        logger.info(
```
Comment (psathyan) [inline ceph/ceph_admin/host.py:88]: Do we need this log statement? There is a log statement in `exec_command` already.

File: ceph/ceph_admin/iscsi.py [hunk 7]
```diff
@@ -1,77 +1,50 @@
...
-        )
+        pos_args = config.pop("pos_args")
 
-        Apply.apply(
-            self,
-            role=self.__ROLE,
-            command=cmd,
-            placements=host_placement,
-        )
+        if isinstance(pos_args[-1], list):
```
Comment (psathyan) [inline ceph/ceph_admin/iscsi.py:45]: Do we need to make this dynamic? That is pull the `CephNode` for the node names and then use the associated IP address.

Maybe we can take it up as an RFE.

File: ceph/ceph_admin/shell.py [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+        out, err = self.installer.exec_command(
+            sudo=True,
+            cmd=" ".join(cmd),
+            timeout=timeout,
+            check_ec=check_status,
+        )
+
+        out = out.read().decode().strip()
+        err = err.read().decode().strip()
+        LOG.debug(out)
```
Comment (psathyan) [inline ceph/ceph_admin/shell.py:43]: Do we need this statement? I believe `exec_command` already adds a statement at `info` level

**Review by psathyan — COMMENTED** [review]
Thanks for doing all the heavy lifting.

Minor comments.

File: ceph/ceph_admin/bootstrap.py (hunk 4 above)
Comment (sunilkumarn417) [reply ceph/ceph_admin/bootstrap.py:70]: However currently node cannot be other than installer node.

File: ceph/ceph_admin/host.py (hunk 6 above)
Comment (sunilkumarn417) [reply ceph/ceph_admin/host.py:88]: we can keep this, just to make sure about config provided

File: ceph/ceph_admin/shell.py (hunk 8 above)
Comment (sunilkumarn417) [reply ceph/ceph_admin/shell.py:43]: I think this is needed, to validate CLI response(for debug purpose).
However currently all runs going with info log-level, which doesn't include debug log records.

File: ceph/ceph_admin/apply.py (hunk 2 above)
Comment (sunilkumarn417) [reply ceph/ceph_admin/apply.py:7]: added

File: ceph/ceph_admin/apply.py (hunk 1 above)
Comment (sunilkumarn417) [reply ceph/ceph_admin/apply.py]: added `OrchApplyServiceFailure`.

**Review by psathyan — APPROVED** [review]
Looks good to me

**Pull request merged**